lib/mongo_client.js

```diff
@@ -241,28 +243,35 @@ MongoClient.prototype.close = function(force, callback) {
     force = false;
   }
 
-  const client = this;
   return maybePromise(this, callback, cb => {
     const completeClose = err => {
-      client.emit('close', client);
+      this.emit('close', this);
 
-      if (!(client.topology instanceof NativeTopology)) {
-        for (const item of client.s.dbCache) {
-          item[1].emit('close', client);
+      if (!(this.topology instanceof NativeTopology)) {
+        for (const item of this.s.dbCache) {
+          item[1].emit('close', this);
         }
       }
 
-      client.removeAllListeners('close');
+      this.removeAllListeners('close');
+
+      // clear out references to old topology
+      this.topology = undefined;
+      this.s.dbCache = new Map();
```

Comment:
the old `dbCache` still references the old topology, so it is manually cleared out here in addition to `this.topology`